infrastructure/smart-contracts/contracts/Achievements.sol

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/access/IAccessControl.sol";
+
+/**
+ * @title Achievements Contract
+ * @dev An ERC1155 contract that represents game achievements.
+ */
+contract Achievements is ERC1155 {
+
+    //constants
+    bytes32 public constant MINT_ROLE = keccak256("MINT_ROLE");
+
+    //state
+    IAccessControl public factoryAddress; //factory proxy to check access control
+    mapping(uint256 => bool) public definedTokens; //tokenId => exists
+    uint256[] private _tokenIdList; //array of token ids that exist (ordered by mint order)
+    uint256 public tokenIdCount; //counter for unique token ids that exist
+    uint256 public totalSupply; //counter for total tokens minted on this contract
+    mapping(uint256 => uint256) public totalSupplyById; //tokenId => tokenCount
+
+    /**
+     * @dev Achievements constructor.
+     * @param _factoryAddress Address of the factory contract that produced this contract.
+     * @param _uri URI string pointing to metadata (e.g. https://metadata.xai.games/id.json)
+     */
+    constructor(address _factoryAddress, string memory _uri) ERC1155(_uri) {
+        factoryAddress = IAccessControl(_factoryAddress);
+    }
+
+    /**
+     * @dev Restricts caller to MINT_ROLE holder in factory contract.
+     */
+    modifier onlyMintRole() {
+        require(factoryAddress.hasRole(MINT_ROLE, address(msg.sender)), "caller does not have MINT_ROLE");
+        _;
+    }
+
```

Comment:
This will do an external request to the factory on every call.
I don't understand why this is needed. We expect only the factory contract itself to call the restricted functions right ?
We should only have to check `msg.sender == factoryAddress` in this contract.